app/models/detector/citation.rb

```diff
@@ -0,0 +1,156 @@
+  # Detector::Citation attempts to identify citations based on the prevalence of individual sub-patterns. It is not
+  # targeted at a particular citation format, but was designed based on characteristics of five formats: APA, MLA,
+  # Chicago, Terabian, and IEEE.
+  #
+  # It receives a Term object, which is parsed in various ways en route to calculating a final score. Terms with a
+  # higher score are more citation-like, while a score of 0 indicates a Term that has no hallmarks of being a citation.
+  # Terms whose score is higher than the REQUIRED_SCORE value can be registered as a Detection.
+  class Citation
+    attr_reader :score, :subpatterns, :summary
+
+    # Citation patterns are regular expressions which attempt to identify structures that are part of many citations.
+    # This object is used as part of the pattern_checker method. Some of these patterns may get promoted to the Detector
+    # model if they prove useful beyond a Citation context.
+    CITATION_PATTERNS = {
+      apa_volume_issue: /\d+\(\d+\)/,
+      no: /no\.\s\d+/,
+      pages: /\d+-+\d+/,
+      pp: /pp\.\s\d+/,
+      vol: /vol\.\s\d+/,
+      year_parens: /\(\d{4}\)/,
+      brackets: /\[.*?\]/,
+      lastnames: /[A-Z][a-z]+[.,]/,
+      quotes: /&quot;.*?&quot;/
+    }.freeze
+
+    # The required score value is the threshold needed for a Term to be officially recorded with a Detection.
+    REQUIRED_SCORE = 6
+
+    # Summary thresholds are used by the calculate_score method. This class counts the number of occurrences of specific
+    # characters in the @summary instance variable. The thresholds here determine whether any of those counts are high
+    # enough to contribute to the Term's citation score.
+    SUMMARY_THRESHOLDS = {
+      characters: 25,
+      colons: 2,
+      commas: 2,
+      periods: 2,
+      semicolons: 2,
+      words: 5
+    }.freeze
+
+    # The initializer handles the parsing of a Term object, and subsequent population of the @subpatterns, @summary,
+    # and @score instance variables. @subpatterns contains all the citation components which have been flagged by the
+    # CITATION_PATTERNS hash. @summary contains counts of how often certain characters or words appear in the Term.
+    # Finally, the @score value is a summary of how many elements in the subpatterns or summary report were detected.
+    #
+    # @note This method can be called directly via Detector::Citation.new(Term). It is also called indirectly via the
+    #       Detector::Citation.record(Term) instance method. This method can be called directly when a Detection is not
+    #       desired.
+    def initialize(term)
+      @subpatterns = {}
+      @summary = {}
+      pattern_checker(term.phrase)
+      summarize(term.phrase)
+      @score = calculate_score
+    end
+
+    # The record method first runs all of the parsers by running the initialize method. If the resulting score is higher
+    # than the REQUIRED_SCORE value, then a Detection is registered.
```

Comment:
It feels a bit odd to me that the record method is where the decision as to whether the input is a citation or not is made.

I wonder if adding a boolean method "citation?" would be valuable. That would give us more flexibility to report that information to UIs that might want it in the future (I'm fine with not returning anything in GraphQL just yet as we learn more about this Detector's usefulness, but having something in GraphQL that indicated we think it was a citation does feel potentially useful)

```ruby
def citation?
  @score >= REQUIRED_SCORE
end
```

Then record could leverage that rather than doing that calculation itself.

```ruby
return unless cit.citation?
```

Reply:
Yeah, this feels like a good method. I'll revise the PR to add this, at least. What do you think about `detection?` as a name, though? That might be shareable across models?

Reply:
`@detection` is what I'm toying with for all pattern matches detected as an instance variable. The expectation would be that it would be an array of name/value pairs.

Having a boolean `detection?` as a convenience method (which on some models would just be the presence of anything in `detection` but for this could be a subset that match the score threshold) does seem like a good pattern.

Reply:
I've pushed a commit and tests to add this, and converted the `record` method to consume it.